shell.nix

```diff
@@ -0,0 +1,140 @@
+  # demonstrate how you could make the behavior a little configurable.
+  #
+  # By default, it will run:
+  # - "all" dev build (--argstr dev minimal or none to disable)
+  # - the "smoke" tests (--argstr test osh-all or none to disable)
+  # - and clean on exit (ex: --argstr cleanup none to disable)
+  build_oil = if dev == "none" then
+    "echo Skipping Oil build."
+  else ''
+    echo Building \'${dev}\' Oil.
+    "$PWD/build/dev.sh" ${dev}
+  '';
+  test_oil = if test == "none" then
+    "echo Skipping Oil tests."
+  else ''
+    echo Running Oil \'${test}\' tests.
+    "$PWD/test/spec.sh" ${test}
+  '';
+  cleanup_oil = if cleanup == "none" then ''
+    echo "Note: nix-shell will *NOT* clean up Oil dev build on exit!"
+    trap 'echo "NOT cleaning up Oil dev build."' EXIT
+  '' else ''
+    echo "Note: nix-shell will clean up the dev build on exit."
+    trap 'echo "Cleaning up Oil dev build."; "$PWD/build/dev.sh" clean' EXIT
+  '';
+in with pkgs;
+
+let
+  # Most of the items you listed in #513 are here now. I'm not sure what the
+  # remaining items here mean, so I'm not sure if they're covered.
+  #
+  # static analysis
+  #   mypy library for mycpp
+  # benchmarks
+  #   ocaml configure, etc. This is just source though.
+  # C deps
+  #   python headers for bootstrapping
+  # big one: Clang for ASAN and other sanitizers (even though GCC has some)
+  #   Clang for coverage too
+
+  # nixpkgs: busybox linux only; no smoosh
+  # could append something like: ++ lib.optionals stdenv.isLinux [ busybox ]
+  spec_tests = [ bash dash mksh zsh ];
+
+  static_analysis = [
+    mypy # This is the Python 3 version
+    (python2.withPackages (ps: with ps; [ flake8 pyannotate ]))
+    # python3Packages.black # wink wink :)
+  ];
+
+  binary = [ re2c ];
+  doctools = [ cmark ];
+  c_deps = [ readline ];
+
+  shell_deps = [
+    gawk
+    time
+    # additional command dependencies I found as I went
+    # guessing they go here...
+    # run with nix-shell --pure to make missing deps easier to find!
+    file
+    git
+    hostname
+    which
+  ];
+
+  nix_deps = [
+    nixfmt # `nixfmt shell.nix` to format in place
+  ];
+
+  # Create a shell with packages we need.
+in mkShell rec {
+
+  buildInputs = c_deps ++ binary ++ spec_tests ++ static_analysis ++ doctools
+    ++ shell_deps ++ nix_deps ++ doctools;
+
+  # Not sure if this is "right" (for nix, other platforms, etc.)
+  # doctools/cmark.py hardcoded /usr/local/lib/libcmark.so, and it looks
+  # like Nix has as much trouble with load_library as you have. For a
+  # "build" I think we'd use a patchPhase to replace the hard path
+  # in cmark.py with the correct one. Since we can't patch the source here
+  # I'm hacking an env in here and in cmark.py. Hopefully others will
+  # weigh in if there's a better way to handle this.
+  #
+  # Note: Nix automatically adds identifiers declared here to the environment!
+  _NIX_SHELL_LIBCMARK = if stdenv.isDarwin then
+    "${cmark}/lib/libcmark.dylib"
```

Comment:
note `stdenv.hostPlatform.extensions.sharedLibrary` exists for shared objects

Reply:
I guess this means this statement could be simplified to the below?
```
_NIX_SHELL_LIBCMARK = "${cmark}/lib/libcmark${stdenv.hostPlatform.extensions.sharedLibrary}"
```

Reply:
Correct

Reply:
Thanks for pointing it out